### PR TITLE
JIT: Type fat pointer locals precisely, and avoid unnecessary normalization in inlining

### LINKED ISSUE #99798: Assertion failed '!"Unexpected tree op after call marked as tailcall"'
Currently failing in main. To repro:

```
build clr.aot+libs.native+libs.sfx -rc Checked -lc Release
src\tests\build.cmd nativeaot checked tree managed
```

Then find find the failing ilc.exe command line in the build log if you need it under debugger (look for "exited with code").

JitDump:

[dump.txt](https://github.com/dotnet/runtime/files/14611732/dump.txt)

Cc @jakobbotsch @EgorBo

### PR DESCRIPTION
**Pull request by jakobbotsch**
During call importation, for fat pointer calls we will introduce a local and spill the call to it. This loses track of the small typedness of the value, which can cause inlining to introduce unnecessary normalization casts later. For tailcalls this can cause us to add IR after the call that we do not expect, causing issues like #99798.

Fix the problem by enhancing logic in a few places:
- Make the local created for these fat pointer calls small typed like regular small normalize-on-store locals
- Enhance `fgCastNeeded` to take into account the small-typedness of these locals (like `IntegralRange::ForNode`)

Some regressions because we remove unnecessary casts which makes this early-out in forward sub kick in:
https://github.com/dotnet/runtime/blob/b94364e90056c371e17e057f352f9f9e9c55511c/src/coreclr/jit/forwardsub.cpp#L724-L743

Fix #99798

### CODE REVIEW

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @EgorBo 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=604583&view=ms.vss-build-web.run-extensions-tab)

**Review by EgorBo — APPROVED** [review]
Do you know why the assert started failing recently?

File: src/coreclr/jit/importercalls.cpp [hunk 1]
```diff
@@ -1406,11 +1406,19 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
...
+                        // Keep the information about small typedness to avoid
+                        // inserting unnecessary casts around normalization.
+                        if (call->IsCall() && varTypeIsSmall(call->AsCall()->gtReturnType))
+                        {
+                            assert(call->AsCall()->NormalizesSmallTypesOnReturn());
+                            varDsc->lvType = call->AsCall()->gtReturnType;
+                        }
 
+                        // TODO-Bug: CHECK_SPILL_NONE here looks wrong.
                         impStoreTemp(calliSlot, call, CHECK_SPILL_NONE);
```
Comment (EgorBo) [inline src/coreclr/jit/importercalls.cpp:1422]: why not fix it then?

**Comment by jakobbotsch** [comment]
> Do you know why the assert started failing recently?

It's exposed by #99265. The callee is a shared generic in this case.

File: src/coreclr/jit/importercalls.cpp (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/importercalls.cpp:1422]: I opened #99808 about fixing it. I mainly think we should separate the diffs + create a regression test for it.

**Comment by EgorBo** [comment]
> > Do you know why the assert started failing recently?
> 
> It's exposed by #99265. The callee is a shared generic in this case.

Weird, I had many CI runs in my PR and didn't see it there, but it's possible of course

**Comment by jakobbotsch** [comment]
> Weird, I had many CI runs in my PR and didn't see it there, but it's possible of course

It did fail in the runtime-nativeaot-outerloop run on that PR: https://dev.azure.com/dnceng-public/public/_build/results?buildId=596527&view=results (you can see the assert in the console log for "windows-x64 Checked NativeAOT_Pri0" there).

**Pull request merged**